### PR TITLE
Use CuPy's `UnownedMemory` for RMM-backed allocations

### PR DESCRIPTION
**Pull request by jakirkham**
Instead of rolling our own class to hold RMM-backed allocations, use CuPy's `UnownedMemory` object. This behaves the same as our hand-rolled class, but saves us the maintenance.

### CODE REVIEW

File: python/rmm/rmm.py [hunk 1]
```diff
@@ -268,7 +252,11 @@ def rmm_cupy_allocator(nbytes):
     if cupy is None:
         raise ModuleNotFoundError("No module named 'cupy'")
 
-    return cupy.cuda.memory.MemoryPointer(RMMCuPyMemory(nbytes), 0)
+    buf = librmm.device_buffer.DeviceBuffer(size=nbytes)
+    mem = cupy.cuda.UnownedMemory(ptr=buf.ptr, size=buf.size, owner=buf)
```
Comment (madsbk) [inline python/rmm/rmm.py]: To support empty arrays, we should set the device id.
```python
dev_id = -1 if buf.ptr else cupy.cuda.device.get_device_id()
mem = cupy.cuda.UnownedMemory(ptr=buf.ptr, size=buf.size, owner=buf, device_id=dev_id)
```
But let's only set it on empty arrays so we don't impose wrong IDs on non-null pointers

Comment (jakirkham) [reply python/rmm/rmm.py]: Thanks Mads! Should be fixed in the latest commit 🙂

**Comment by jakirkham** [comment]
Submitted PR ( https://github.com/cupy/cupy/pull/3086 ) upstream to add `UnownedMemory` to the API docs.

File: python/rmm/tests/test_rmm.py [hunk 2]
```diff
@@ -250,16 +250,16 @@ def test_rmm_cupy_allocator():
     m = rmm.rmm_cupy_allocator(42)
     assert m.mem.size == 42
     assert m.mem.ptr != 0
-    assert m.mem.rmm_array is not None
+    assert isintance(a.data.mem._owner, rmm.DeviceBuffer)
```
Comment (kkraus14) [inline python/rmm/tests/test_rmm.py]:
```suggestion
    assert isinstance(a.data.mem._owner, rmm.DeviceBuffer)
```

File: python/rmm/tests/test_rmm.py [hunk 3]
```diff
@@ -250,16 +250,16 @@ def test_rmm_cupy_allocator():
...
 
     m = rmm.rmm_cupy_allocator(0)
     assert m.mem.size == 0
     assert m.mem.ptr == 0
-    assert m.mem.rmm_array is None
+    assert isintance(a.data.mem._owner, rmm.DeviceBuffer)
```
Comment (kkraus14) [inline python/rmm/tests/test_rmm.py]:
```suggestion
    assert isinstance(a.data.mem._owner, rmm.DeviceBuffer)
```

File: python/rmm/tests/test_rmm.py [hunk 4]
```diff
@@ -250,16 +250,16 @@ def test_rmm_cupy_allocator():
...
 
     cupy.cuda.set_allocator(rmm.rmm_cupy_allocator)
     a = cupy.arange(10)
-    assert hasattr(a.data.mem, "rmm_array")
+    assert isintance(a.data.mem._owner, rmm.DeviceBuffer)
```
Comment (kkraus14) [inline python/rmm/tests/test_rmm.py]:
```suggestion
    assert isinstance(a.data.mem._owner, rmm.DeviceBuffer)
```

**Comment by jakirkham** [comment]
Thanks Keith! 😄

**Pull request merged**